pkg/build/graph/analysis/bc.go

```diff
@@ -0,0 +1,45 @@
+package analysis
+
+import (
+	"fmt"
+	"strings"
+
+	osgraph "github.com/openshift/origin/pkg/api/graph"
+	buildedges "github.com/openshift/origin/pkg/build/graph"
+	buildgraph "github.com/openshift/origin/pkg/build/graph/nodes"
+	imageedges "github.com/openshift/origin/pkg/image/graph"
+	imagegraph "github.com/openshift/origin/pkg/image/graph/nodes"
+)
+
+const (
+	MissingRequiredRegistryWarning = "MissingRequiredRegistry"
+)
+
+// FindUnpushableBuildConfigs checks all build configs that will output to an IST backed by an ImageStream and checks to make sure their builds can push.
+func FindUnpushableBuildConfigs(g osgraph.Graph) []osgraph.Marker {
+	bcNames := []string{}
+
+bc:
+	for _, bcNode := range g.NodesByKind(buildgraph.BuildConfigNodeKind) {
+		for _, istNode := range g.SuccessorNodesByEdgeKind(bcNode, buildedges.BuildOutputEdgeKind) {
+			for _, uncastImageStreamNode := range g.SuccessorNodesByEdgeKind(istNode, imageedges.ReferencedImageStreamGraphEdgeKind) {
+				imageStreamNode := uncastImageStreamNode.(*imagegraph.ImageStreamNode)
+				if len(imageStreamNode.Status.DockerImageRepository) == 0 {
+					bcNames = append(bcNames, bcNode.(*buildgraph.BuildConfigNode).ResourceString())
+					continue bc
+				}
+			}
+		}
+	}
+
+	if len(bcNames) == 0 {
+		return []osgraph.Marker{}
+	}
+
+	return []osgraph.Marker{{
```

Comment:
Keep - this is a specific use case we've detected (i don't even think you
need to enumerate the individual items, but it's fine to keep).  There is
no need to hit the user over the head.

On Fri, Jul 17, 2015 at 3:08 PM, Dan Mace <notifications@github.com> wrote:

> In pkg/build/graph/analysis/bc.go
> <https://github.com/openshift/origin/pull/3763#discussion_r34921356>:
>
> > +		for _, istNode := range g.SuccessorNodesByEdgeKind(bcNode, buildedges.BuildOutputEdgeKind) {
> > +			for _, uncastImageStreamNode := range g.SuccessorNodesByEdgeKind(istNode, imageedges.ReferencedImageStreamGraphEdgeKind) {
> > +				imageStreamNode := uncastImageStreamNode.(*imagegraph.ImageStreamNode)
> > +				if len(imageStreamNode.Status.DockerImageRepository) == 0 {
> > +					bcNames = append(bcNames, bcNode.(*buildgraph.BuildConfigNode).ResourceString())
> > +					continue bc
> > +				}
> > +			}
> > +		}
> > +	}
> > +
> > +	if len(bcNames) == 0 {
> > +		return []osgraph.Marker{}
> > +	}
> > +
> > +	return []osgraph.Marker{{
>
> I'd like discrete markers. If a caller wants to collapse them based on the
> key and the related nodes, that's always an option. By compressing the
> markers here you lose information (the actual IST nodes related to the BCs).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/3763/files#r34921356>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift